### ISSUE TITLE
Rationale behind CORS hook setting blank headers

### ISSUE THREAD

**Issue by Yn5an3** [balderdashy/sails#3593]
I'm wondering why does Sails set up blank headers as a part of the CORS hooks. While it does not seem like a problem, it conflicts with any other header set by a reverse proxy, and duplicate headers may end being sent to client unless those headers are cleared first.

First, let me put an example. I'm trying to create something similar to a widget service. In order to achieve that, a piece of JS code loads a URL via Ajax. Due to the giantic amount of traffic the widget would generate, it's mandatory for us to set up a CDN to cache the rendered widgets. We use cloudflare to cache every url of the form widget/render/:id. The problem happens when you want to set '*' as the allowed origin, in order to have those headers cached. 

One possibility would be for Sails to send the '*' itself, other than replacing the widcard by the domain where the request originates, so the CDN, or any reverse proxys for that matter, were able to cache it properly. The other would be to prevent sails from sending empty Access-Control headers, so any webserver in front of Node would be setting the headers. 

The last observation would be that, when CORS is enabled setting Access-Control-Allow-Origin to something different than '*', shouldn't the Vary: Origin header be present in order to allow reverse proxys to cache different versions of the content?

**Comment by sgress454** [comment]
Hi @Yn5an3,

Good points.  To answer the general question, the sending of blank headers vs. no headers started as a workaround for an issue we had with clearing existing headers in Express.  It's probably no longer an issue, so as long as the tests still pass, I'd be fine with using `.removeHeader(headerName)` instead of using `.set(headerName, '')`.

As far as just sending `*`, at the time we were developing the hook, that seemed like a security flaw, but the more I think about it it seems that if your security rests on people not _knowing_ that your API is actually open to the world, then you've got more problems than we can solve by hiding that fact for you.  So changing it to `Access-Control-Allow-Origin: *` would be fine (provided there was a test).

Finally, re: using `Vary: Origin`, it just wasn't on our radar when the CORS hook was first developed, but it's an excellent observation.  If you'd care to [write up a proposal](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md#v-proposing-features-and-enhancements) for adding the header to Sails requests under the right circumstances, it's definitely something we could add to the roadmap.  I'd prefer it to be toggleable (although it could be on by default) just because it does reveal something (however small) about the server configuration, so some folks might prefer to keep it hidden if they don't need it.

**Comment by Yn5an3** [comment]
I'll try to write proposal then. If I have questions, I'll just drop by the IRC channel.

**Comment by Yn5an3** [comment]
I forgot to update on this. Many of the tests do break when replacing the `.set(headerName, '')` for `.removeHeader(headerName)`, as they expect those headers to be present in the responses. I think it's better to leave things as they are. It's a small change that would cost us nothing to implement, but the impact in the test suite/hook is quite big. 

My second concern is that I don't know if the current implementation's behaviour was intentional or not. As you've said, it could be that you considered doing it in a different way would be a security flaw, or would expose information that you didn't want to reveal. For example is not uncommon, if no origin header is sent, or no match is found for the header, to stop CORS processing, [as per the specs](https://www.w3.org/TR/cors/), point 6.1. Sails does something similar, when the security level is increased, but that controls whether the request is processed or not, and not the access to the representation of the response. I'd like to know your opinion on the matter. 

Thanks for your patience.

**Comment by sgress454** [comment]
@Yn5an3 By default, Sails follows the CORS spec precisely: if there is no origin header then it considers the request as if it was from the same origin and allows it; the assumption is that the request didn't come from a browser, so the blank `access-control-allow-origin` header will have no effect.  Turning the security setting up to "very high" makes things more restrictive; you _must_ have a matching origin in order for the request to be processed at all, otherwise you'll get a 403.

The issue with cacheing seems to me to warrant updating the way we send the headers; if that means changing those tests to check for the non-existence of the header instead of for an empty string, then so be it!

**Comment by Yn5an3** [comment]
> 6.1.2 - If the value of the Origin header is not a case-sensitive match for any of the values in list of origins **do not set any additional headers** and terminate this set of steps.

@sgress454.  Then should I assume this is correct? 
[https://github.com/balderdashy/sails/blob/master/lib/hooks/cors/to-prepare-send-headers.js#L74
](https://github.com/balderdashy/sails/blob/master/lib/hooks/cors/to-prepare-send-headers.js#L74
)
I think, if no origin is deemed valid (no match is found), that line is not following the specs. The same happens to lines 77-87; those should be skipped... I think.  

It's relevant in the event tests are changed. The behavior should be the same if no Origin header is present or no match is found. That line would make it inconsistent if we switch to `.removeHeader(headerName)`.

Maybe I'm just overthinking this... 

Anyways. Would something like [this paste](http://pastebin.com/emM9MeAz) be ok? It's just a quick **untested** attemp that condenses all the changes. Only for illustrative purposes, I know this would need to be split in several commits and I'd need to make the formal proposal first in order to include the Vary header, etc.

**Comment by mikermcneil** [comment]
@Yn5an3 If you wouldn't mind opening a proposal for this, that'd be awesome (it'd give us a place to consolidate any documentation links that would need to change as well).  thanks for the help!

(I'm going to flag this as "bug" to keep this open in the mean time, but feel free to close this and link to it from your proposal when ready)

**Comment by Yn5an3** [comment]
@mikermcneil Ok, will do. Thanks for commenting on the issue! I was afraid of doing something wrong, actually, that's why I posted all those questions. 

And thanks for your awesome work in this project.

**Comment by sgress454** [comment]
@mikermcneil Just FYI Botsy will ignore issues with `bug`, `inconsistency` or `performance` labels now.

**Comment by mikermcneil** [comment]
@sgress454 oh awesome!

**Comment by sgress454** [comment]
Closing this as the proposal has been added to the backlog.  Thanks again @Yn5an3!
